src/Microsoft.OData.Core/UriParser/Resolver/CaseInsensitiveSchemaElementsCache.cs

```diff
@@ -0,0 +1,136 @@
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.OData.Edm
+{
+    /// <summary>
+    /// Cache used to store schema elements using case-normalized names
+    /// to speed up case-insensitive model lookups. The cache is populated
+    /// up front so that if an item is not found in the cache, we can assume
+    /// it doesn't exist in the model. For this reason, it's important that
+    /// the model be immutable.
+    /// </summary>
+    internal sealed class CaseInsensitiveSchemaElementsCache
+    {
+        private static readonly List<IEdmSchemaElement> emptyList = new List<IEdmSchemaElement>();
+        // This cache is meant to be populate up front and remain read-only
+        // after that. Therefore, it doesn't need
+        // to be a concurrent dictionary.
+        private readonly Dictionary<string, List<IEdmSchemaElement>> cache;
+
+        /// <summary>
+        /// Builds a case-insensitive cache of schema elements from
+        /// the specified <paramref name="model"/>.
+        /// </summary>
+        /// <param name="model">The model whose schema elements to cache. This model should be immutable. See <see cref="ExtensionMethods.MarkAsImmutable(IEdmModel)"/>.</param>
+        public CaseInsensitiveSchemaElementsCache(IEdmModel model)
+        {
+            Dictionary<string, List<IEdmSchemaElement>> cache = new Dictionary<string, List<IEdmSchemaElement>>(StringComparer.OrdinalIgnoreCase);
+
+            PopulateSchemaElements(model, cache);
+
+            foreach (IEdmModel referencedModel in model.ReferencedModels)
+            {
+                PopulateSchemaElements(referencedModel, cache);
+            }
+
+            this.cache = cache;
+        }
+
+        /// <summary>
+        /// Find all schema elements that match the specified <paramref name="qualifiedName"/>.
+        /// </summary>
+        /// <param name="qualifiedName">The case-insensitive fully qualified name to match.</param>
+        /// <returns>The list of elements that matched the <paramref name="qualifiedName"/>.</returns>
+        public IReadOnlyList<IEdmSchemaElement> FindElements(string qualifiedName)
+        {
+            if (cache.TryGetValue(qualifiedName, out List<IEdmSchemaElement> results))
+            {
+                return results;
+            }
+
+            return emptyList;
+        }
+
+        /// <summary>
+        /// Find all schema elements of type <typeparamref name="T"/> that match the specified <paramref name="qualifiedName"/>.
+        /// </summary>
+        /// <typeparam name="T">The type of schema elements to match.</typeparam>
+        /// <param name="qualifiedName">The case-insensitive fully qualified name to match.</param>
+        /// <returns>The list of elements of type <typeparamref name="T"/> that matched the <paramref name="qualifiedName"/>.</returns>
+        public IReadOnlyList<T> FindElementsOfType<T>(string qualifiedName) where T : IEdmSchemaElement

```

Comment:
From the code it seems these O(n) searches (**FindElementsOfType**, **FindSingleOfType**) will still happen in AGS, even with the immutable cache? They can't be made O(1)? Or it is not worth it?

Reply:
In this case O(n) would be relative to the number of elements that match the same name, rather than the total number of elements in the schema. If the AGS schema doesn't have schema elements with similar names, then this will be close to O(1) since each list would only have a single entry.  We can probably achieve O(1) if we store elements of different types in different collections, for example one dictionary for schema types, one dictionary for operations, etc. so we don't have to filter by type

Reply:
Sounds good. And if these lists are expected to be small (e.g., less than ~4 elements) for most cases, List search can actually beat Dictionary.

@syprieur this should be small lists in AGS scenario?

Reply:
For smaller models, with no name collisions, I think the extra dictionaries might cost more overhead. However, the concern is with the `List<T>` allocations in `FindElementsOfType<T>`. I try to optimize the case where only one item is needed with `FindSingleOfType<T>` to be allocation-free in the common case. But if `FindElementsOfType<T>` is still on the hot path, then it may be worthwhile eliminating the `new List<T>`.

Reply:
I've created an experiment based on multiple caches (1 cache for schema types, 1 cache for operations, 1 cache for vocabulary terms and 1 cache for every other schema element). So instead of calling `FindElementsOfType<IEdmOperation>`, I'd call `FindOperations` which looks up the cache that stores `IEdmOperation` instances. So, I don't need to create a new `List<T>` each time. On my benchmark (which may or may not not be representative of AGS) memory allocations went slightly down (from 67.41MB to 67.1MB).

Reply:
Yes, these Lists will be small for AGS: we enforce FQN unicity on our models (with 1-2 exceptions for legacy APIs), so as Clément stated, we'll only have more than one element for operations overloads.